### DISCUSSION TITLE
Teensy4.1 Sporadically Resets in the middle of a job

### DISCUSSION THREAD

**Discussion by jerseyguy1996** [terjeio/grblHAL#233] [Q&A]
I'm running a Teensy4.1 on a K40 laser engraver.  I have this sporadic issue that pops up where right in the middle of engraving an image it will reset causing it to lose its place.  The worst part is that once it starts running again, rather than being locked it will immediately start engraving again with the next instruction even though it's no longer in the right position.  This is the error:

`ALARM:3
Reset while in motion. Grbl cannot guarantee position. Lost steps are likely. Re-homing is highly recommended.
ok
GrblHAL 1.1f ['$' for help]`

It's seems to be completely random as to when it does it.  Sometimes it will complete three or four jobs without any issues.
Does anyone have any ideas why it might be resetting in the middle of a job?

**Comment by phil-barrett** [comment]
I am not familiar with the K40 setup.  How is the BOB wired in?

**Reply by phil-barrett** [reply]
Also, is this USB or Ethernet connected?

**Reply by jerseyguy1996** [reply]
I'm not gonna lie....it's pretty messy right now because its all just breadboarded with jumper wires.  I'm concerned that it's not the wiring though because I think it was doing it when I was running gcode through a bare Teensy with nothing wired up to it.  Basically it's cranking along just fine and then suddenly I would see the Pin13 LED come on (the initialized state as it is also the laser enable pin which is triggered low), and then go off again as it would resume the code after it re-booted.  Just to verify that it isn't the wiring I'll run some jobs on the bare board again and see if I get the error with nothing connected to it.

**Comment by phil-barrett** [comment]
There are a lot of possibilities. EMI is always a potential issue. What is source of your firmware?  Did you build or use a prebuilt one?

Would you mind sharing the gcode file that gets a reset and I will test it here.

**Reply by jerseyguy1996** [reply]
I'm going to the makerspace in a bit to pull the controller and I'll run some gcode through it to see if the alarm occurs when it isn't connected to anything.  The firmware was compiled from the Grbl Hal master branch with some minor config changes compiled using the Arduino IDE.  I'm curious, are there any conditions that are supposed to trigger a software reset, such as hitting an endstop?  I don't have anything connected to the endstop pins and I inverted the logic to trigger on low.  If the endstop triggers due to EMI, does that trigger a software reset?

**Reply by terjeio** [reply]
> If the endstop triggers due to EMI, does that trigger a software reset?

If hard limits is enabled then yes. The Halt (or reset/e-stop) input may also trigger a software reset.

A K40 is likely a very noisy machine regading EMI (my CO2 laser surely is), limit or control signal inputs should definitely be normally closed (NC) switches and if not used they should be shorted to ground (with a short piece wire) instead of inverting the inputs.

**Reply by jerseyguy1996** [reply]
Looks like I don't have hard limits enabled but I do have everything inverted (got lazy and didn't want to have to ground all of the signal pins).  I think I need to (re) - comment this out and ground everything.  Feeling like that is probably the issue.  Gonna guess it's the RESET_PIN triggering it (I didn't even know there was a reset pin).

`#define INVERT_CONTROL_PIN_MASK SIGNALS_BITMASK // Default disabled. Uncomment to enable.`[](url)

**Reply by phil-barrett** [reply]
Terje, I thought the reset pin did not cause an actual reset in the T4.1 version. 

If this is on a solderless breadboard, there would be a number of ways for EMI to get in and reset the processor.

**Reply by jerseyguy1996** [reply]
Well it seems like it's not actually a "reset" that I'm seeing.  After the glitch I was pressing the "Stop" button on Lightburn which was triggering a reset, so the Alarm 3 was a result of me stopping the job after the glitch happened.  I'm still testing things but I think the glitch is actually a result of my computer.  I have my screen set to turn off after 5 minutes and it seems that the glitch happens when the screen turns off.  I'm not sure why that would affect the Teensy though as the computer itself isn't going to sleep, just the screen.

Just to describe exactly what is happening (and there is no error message or alarm being generated when this happens) is the machine will be engraving as normal, it will suddenly stop in the middle of the engraving, when it stops it will disable the laser, then after a split second it will enable the laser and start engraving again where (I think) it left off but it's position will be shifted over.  It seems to me that even if there was a momentary glitch in communication with the computer, it should have enough instructions in the buffer to keep going and the fact that it's losing its position has got me befuddled.  I'll try to get a video of it happening.  If it turns out it's just a result of my screen turning off then it will be an easy fix.  Hopefully that is all it is.

**Reply by jerseyguy1996** [reply]
@phil-barrett  @terjeio Well sadly it wasn't an issue with my computer because it happens with other computers as well.  I'm thinking it might be a communication issue.  I've made a short video of the glitch.  It occurs almost immediately when the job starts in the attached video, and it occurs again at about the 2:26 mark just before the end of the video.  Have either of you ever seen this before?

https://www.youtube.com/watch?v=SUHePaLqoUw&feature=youtu.be

**Comment by phil-barrett** [comment]
I have not seen this kind of behavior.  Is it still the Alarm mentioned in your first post?  

Could you take a few minutes to give us some information:
- your hardware set up. How it is connected to the K40. Maybe some pictures. 
- the GCode for the program that shows the issue
- what sender you are using
- your grblHAL settings ($$).
- the results of sending $i to the teensy.
- If you are you building, from the Main or Test branch?

That will help a lot in getting to the bottom of this.

**Reply by jerseyguy1996** [reply]
The alarm wasn't the problem and was merely triggered as a result of the Stop command that I was issuing after the glitch occurred.

Regarding your questions:

1.  It's a bit of a messy set up because i've temporarily wired up an adaptor using a protoboard so that I can plug it in where the old Arduino Nano use to go.  I'm using 74AHCT125's for level shifting.  However, it does it even when it's just the bare Teensy with nothing attached so I don't think it's an issue with the install. I will say that I'm using a pretty cheap USB cable.  I've found it difficult to find a well shielded  cable (with ferrite clamps) in the micro usb style.  Will an ethernet connection be a better option for communication issues?
2. The g-code is generated by Lightburn.  It seems to only do this with raster images which I believe Lightburn uses relative positioning to shorten the lengths of each discrete command.
3. Lightburn but it does it with ioSender as well.  In Lightburn it is using the Buffered transfer mode which I believe is designed to keep the serial receive buffer full.
4. I'll get you these.  I think they are pretty much standard with the exception of Lasermode turned on, $14=78, and step direction pins inverted.
5. I'll try this the next time I'm at the makerspace.  What does $i do?
6. Main branch

Here's a question.  If the USB data line receives interference and the controller receives a string of garbage, I assume that it will just ignore that string and then pick back up on the next valid instruction.  I'm wondering if the issue is that it's receiving a block of garbage and then when it starts up again, because it's in relative positioning mode it loses it's place due to the skipped instructions.  Is there a way to see the response from the controller, either 'ok' or 'command not recognized' in the console.  It seems like those responses aren't displayed on any of the gcode senders.

**Reply by phil-barrett** [reply]
1) Given that it seems EMI related, the HW is a big point of attention for solving the problem. If you'd prefer not showing, OK. I think you get what needs to be done to minimize EMI with your setup.  Ethernet uses differential signaling so it has a degree of immunity. My old CBeam router with an Arduino via USB would get errors regularly, none since I switched to Ethernet.
2) it would be worth trying the GCode on other hardware just to eliminate it as an issue. If nothing else, I would add it to my test suite.
4) I just want to eliminate as many possibilities.
5) $i tells us what build you have so we can know where to look if it's a code issue.

grblHAL's response to garbage is somewhat unpredictable. I know that ioSender will indicate OK next to the block when it has been competed or an error number.  "*" means the block has been send but not yet processed. You can see this in the attached screen shot. 

Lightburn seems to give very little feedback from Grbl. For the most amount of information, it would be good if you used ioSender rather than Lightburn.  You can export the GCode from Lightburn via the "Save GCode" button in the Laser tab.

![iosender](https://user-images.githubusercontent.com/16570958/107992647-fbc51080-6f8d-11eb-9bfa-380c3c77413b.png)

**Reply by phil-barrett** [reply]
By the way, in ioSender directory, there is an error codes CSV file - error_codes_en_US.csv - that will give you a bit more information..

**Reply by jerseyguy1996** [reply]
I'm going to put my results of my most recent tests in a series of posts.  First off I added the Teensy4.1 Ethernet Kit sold by PJRC hoping that that would solve my issues.  It's still glitching but in a slightly different way.  Whereas before it would stop for a second and then start up again in the wrong place, now it doesn't stop but it will slide over to the wrong place and keep going as if nothing is wrong.  I have "Aggressive Buffering" and "Keep MDI focus" checked on ioSender.  Interestingly enough, the machine still knows where it is because if I tell it to return to X0Y0 it goes right back to where it is supposed to.  The fact that the file uses all relative positioning is why I think it is just missing an instruction here or there.  I'll post a picture of my wiring below and then a picture of some of my test engravings on the following posts.  I didn't want to show this because it's quite frankly embarrassing, but it is just temporary until I can make a dedicated board for it.  
![50948011996_bb37a2c66f_c](https://user-images.githubusercontent.com/4678592/108003795-d225dc80-6fc1-11eb-85ed-364278def029.jpg)

**Reply by jerseyguy1996** [reply]
This is one of the engravings.  Notice the top right engraving, that it shifted really only twice (well actually 3 times because it shifted on the second line but it's hard to notice) so you figure 2 (or 3) corrupted instructions out of a file that is 151,524 lines long.  I mean I guess that's not bad but it still ruins the engraving....lol.  I'm seriously thinking of just writing a simple pre-processor that will convert all of the relative positioning that Lightburn produces to absolute coordinates.  I know the individual lines of code will be slightly longer meaning more characters to send but given the increased communication speed of the Teensy I don't think it will be a problem.  That way if it gets a corrupted instruction, the worst that will happen is a stray (or missed) line.  Do you think this is a reasonable fix or should I be shooting for zero corrupted lines in a long file such as this?
![50947287853_61e4b145ec_c](https://user-images.githubusercontent.com/4678592/108004073-bb33ba00-6fc2-11eb-9678-b9838c0f2e8a.jpg)

**Reply by jerseyguy1996** [reply]
This is the g-code file for that engraving.
[Cat Engraving.zip](https://github.com/terjeio/grblHAL/files/5984833/Cat.Engraving.zip)

**Reply by karoria** [reply]
> I'm going to put my results of my most recent tests in a series of posts. First off I added the Teensy4.1 Ethernet Kit sold by PJRC hoping that that would solve my issues. It's still glitching but in a slightly different way. Whereas before it would stop for a second and then start up again in the wrong place, now it doesn't stop but it will slide over to the wrong place and keep going as if nothing is wrong. I have "Aggressive Buffering" and "Keep MDI focus" checked on ioSender. Interestingly enough, the machine still knows where it is because if I tell it to return to X0Y0 it goes right back to where it is supposed to. The fact that the file uses all relative positioning is why I think it is just missing an instruction here or there. I'll post a picture of my wiring below and then a picture of some of my test engravings on the following posts. I didn't want to show this because it's quite frankly embarrassing, but it is just temporary until I can make a dedicated board for it.
> ![50948011996_bb37a2c66f_c](https://user-images.githubusercontent.com/4678592/108003795-d225dc80-6fc1-11eb-85ed-364278def029.jpg)

I can still suggest you eliminating that flat ribbon cable between board and ethernet module. FRC ranks 1st in catching EMI. If you can't do more in existing setup, try to make it very shorter just to connection requirements. If possible, wrap aluminium foil (be careful as it can short open solder joints) over that area, that can help.

**Reply by phil-barrett** [reply]
I agree, the ethernet connection is a likely antenna.  If nothing else, fold it over a couple of times.   You could also make a super short version of the cable and park the Magjack board directly on top of the teensy.

But also, add some LPFs to at least the control signals.  Optos would be better. Not trying to sell you a board but my T4.1 Breakout Board would get you there.

And, try setting block size to 512 as that seems to make the problems I am seeing go away.  I ran your cat gcode with 512 and it seemed to run correctly but it wasn't on a laser engraver.

**Reply by jerseyguy1996** [reply]
I can eliminate the flat ribbon cable and just solder the ethernet board directly to teensy.  I'll give that a try as well as changing my block size.  I'm also writing a simple pre-processor to convert all of the relative moves to absolute in these raster files.  That will be a last resort, or at least what I can do until I work out my emi issues.  Currently I'm not using any of the control signals.  I've inverted the signals but I can reverse that and just ground them all if I need to. 
 Eventually I will use them but probably on a future machine.  @phil-barrett You ever consider designing a board that has headers for the Pololu Stepper Drivers?  I'm a big fan of those and tend to design all of my motion control boards around them.  What stepper drivers are you currently using for your machine with your boards?

**Comment by karoria** [comment]
It seems EMI. Floating or not connected input pins are highly sensitive to EMI. Are your inputs filtered for EMI like RC with opto? Is your USB cable (pc to controller) shielded and having shorter length?

**Reply by jerseyguy1996** [reply]
I'm thinking more and more that it is a communication issue.  It's a pretty cheap USB cable and I haven't been able to find a well shielded micro usb cable with ferrite beads clamped on each end.

**Reply by Klippert** [reply]
I've had issues with bad usb cables, and especially a bad usb connector on the actual board (was an arduino). So if you can make sure the connection is not at fault (and switch usb cable) that could potentially fix it.

**Reply by phil-barrett** [reply]
Certainly worth trying a better USB cable.  However, it may be of secondary importance if there are bare inputs that aren't connected to ground or do not have fairly aggressive RC low pass filtering. The reset error message mentioned earlier makes me think EMI rather than garbled communication but that is just a guess.

**Reply by jerseyguy1996** [reply]
> Certainly worth trying a better USB cable. However, it may be of secondary importance if there are bare inputs that aren't connected to ground or do not have fairly aggressive RC low pass filtering. The reset error message mentioned earlier makes me think EMI rather than garbled communication but that is just a guess.

Ya the reset error was a red herring created by Lightburn.  Clicking on the "Stop" button, which is what I was doing each time the glitch would occur was causing the controller reset.  The reset had nothing at all to do with the glitch.  I'm almost positive that it is a communication issue at this point.  I'll reply on the other thread as to what my latest batch of tests concluded.

**Comment by phil-barrett** [comment]
Hi, are you building with a large planning buffer - block buffer size set to 1024? If so, try commenting out the larger buffer definition.  I am about to submit an issue about that.  Looks like it breaks something. I'd getting some odd behavior. It may not be related to your problem but worth trying.

**Reply by jerseyguy1996** [reply]
I put a series of posts up above that shows my hardware setup, some engravings that I tried today, and the gcode file.  I do have the block buffer size set to 1024.  I'll try to uncomment that.  On a raster image though, I could see how planning many moves ahead will be necessary to keep the speed up as I'm guessing the default buffer size is probably not even half an inch of moves given that a complex image may have hundreds of discrete moves on each line.

**Reply by phil-barrett** [reply]
If it is a bug, we will get it fixed.  Large planning buffer is support is important.
